schain/schainwriter.go

```diff
@@ -0,0 +1,151 @@
+	sChainsByBidder, err := BidderToPrebidSChains(schains)
+	if err != nil {
+		return nil, err
+	}
+	writer.sChainsByBidder = sChainsByBidder
+	return writer, nil
+}
+
+// ORTBTwoFiveSChainWriter implements the SChainWriter interface and is used to write the appropriate schain
+// for a particular bidder defined in the ORTB 2.5 multi-schain location (req.ext.prebid.schain) to the
+// ORTB 2.5 location (req.source.ext)
+type ORTBTwoFiveSChainWriter struct {
+	sChainsByBidder map[string]*openrtb_ext.ExtRequestPrebidSChainSChain
+}
+
+// Write selects an schain from the multi-schain ORTB 2.5 location (req.ext.prebid.schains) for the specified bidder
+// and copies it to the ORTB 2.5 location (req.source.ext). If no schain exists for the bidder in the multi-schain
+// location and no wildcard schain exists, the request is not modified.
+func (w ORTBTwoFiveSChainWriter) Write(req *openrtb2.BidRequest, bidder string) {
+	const sChainWildCard = "*"
+	var selectedSChain *openrtb_ext.ExtRequestPrebidSChainSChain
+
+	wildCardSChain := w.sChainsByBidder[sChainWildCard]
+	bidderSChain := w.sChainsByBidder[bidder]
+
+	// source should not be modified
+	if bidderSChain == nil && wildCardSChain == nil {
+		return
+	}
+
+	if bidderSChain != nil {
+		selectedSChain = bidderSChain
+	} else {
+		selectedSChain = wildCardSChain
+	}
+
+	if req.Source == nil {
+		req.Source = &openrtb2.Source{}
+	} else {
+		sourceCopy := *req.Source
+		req.Source = &sourceCopy
+	}
+	schain := openrtb_ext.ExtRequestPrebidSChain{
+		SChain: *selectedSChain,
+	}
+	sourceExt, err := json.Marshal(schain)
+	if err == nil {
+		req.Source.Ext = sourceExt
+	}
+}
+
+// newORTBTwoFourSChainWriter creates an ORTB 2.4 schain writer instance.
+func newORTBTwoFourSChainWriter(reqExt *openrtb_ext.ExtRequest) SChainWriter {
+	return ORTBTwoFourSChainWriter{
+		schain: reqExt.SChain,
+	}
+}
+
+// ORTBTwoFourSChainWriter implements the SChainWriter interface and is used to write an schain
+// defined in the ORTB 2.4 location (req.ext.schain) to the ORTB 2.5 location (req.source.ext)
+type ORTBTwoFourSChainWriter struct {
+	schain *openrtb_ext.ExtRequestPrebidSChainSChain
+}
+
+// Write copies an schain defined in the ORTB 2.4 location (req.ext.schain) to the ORTB 2.5 location (req.source.ext)
+func (w ORTBTwoFourSChainWriter) Write(req *openrtb2.BidRequest, bidder string) {
+	if w.schain == nil {
+		return
+	}
+
+	if req.Source == nil {
+		req.Source = &openrtb2.Source{}
+	} else {
+		sourceCopy := *req.Source
+		req.Source = &sourceCopy
+	}
+
+	schain := openrtb_ext.ExtRequestPrebidSChain{
+		SChain: *w.schain,
+	}
+	sourceExt, err := json.Marshal(schain)
+	if err == nil {
+		req.Source.Ext = sourceExt
+	}
+}
+
+// extPrebidSChainExists checks if an schain exists in the ORTB 2.5 req.ext.prebid.schain location
+func extPrebidSChainExists(reqExt *openrtb_ext.ExtRequest) bool {
+	if reqExt == nil {
+		return false
+	}
+	if reqExt.Prebid.SChains == nil {
+		return false
+	}
+	return true
+}
```

Comment:
This function was written this way instead of a simple one liner of `return reqExt != nil && reqExt.Prebid.SChains != nil` because I think it helps us ensure code coverage of each condition when running the tests with code coverage enabled. The same applies to the two helper functions right after it.
If anyone prefers I use a one liner LMK. I thought the benefits here outweighed the slight increase in LOC.

Reply:
I was going to write a comment about this :)
I only think it's better to check `reqExt.Prebid.SChains != nil`, but I'm also ok with current logic.

Reply:
I think we also need to check if `reqExt` is nil otherwise we will get an error when attempting to access the `Prebid` field on a nil `reqExt`.